OBSDOCS-1359_New and OBSDOCS-817 (partially covered) (#83850)
openshift/openshift-docs: observability/logging/logging-6.0/log6x-configuring-lokistack-otlp.adoc

:_mod-docs-content-type: ASSEMBLY
[id="log6x-configuring-lokistack-otlp"]
= OTLP data ingestion in Loki
include::_attributes/common-attributes.adoc[]
:context: log6x-configuring-lokistack-otlp

toc::[]

> Loki 3.0 introduces an API endpoint using the OpenTelemetry Protocol (OTLP), providing a new way to ingest log entries into Loki. This endpoint complements the standard Push API, which is available in Loki from its initial release. Since OTLP is a standardized format not specifically designed for Loki, it requires additional configuration on Loki's side to map OpenTelemetry's data format to Loki's data model. OTLP lacks concepts such as *stream labels* or *structured metadata*. Instead, OTLP provides metadata about log entries as *attributes*, grouped into three categories:

Reviewer: Do *stream labels* and *structured metadata* really need to be emphasized? Usually we would use italics for emphasis, but should generally avoid (per [repo guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference)):

> Use of underscores for general emphasis is allowed but should only be used very sparingly. Let the writing, instead of font usage, create the emphasis wherever possible.

Italics can be fine for a "first term/instance" style usage, if it's something that's going to be repeated.

Author: Lets use italics here. 
![Screenshot from 2024-11-08 17-29-26](https://github.com/user-attachments/assets/c05d61bc-ffbe-4e2c-b0a0-5134301cc0fc)

* Resource
* Scope
* Log

This allows metadata to be set for multiple entries simultaneously or individually as needed.

include::modules/log6x-configuring-lokistack-otlp-data-ingestion.adoc[leveloffset=+1]

> == Attribute mapping

Reviewer: Needs unique ID/anchor (and any other subsections in this assembly).

In `openshift-logging` mode, custom attribute mappings can be configured globally for all tenants or for individual tenants as needed. When custom mappings are defined, they are appended to the OpenShift defaults. If default recommended labels are not required, they can be disabled in the tenant configuration. For more information, see xref:../../../observability/logging/logging-6.0/log6x-configuring-lokistack-otlp.adoc#customizing-openshift-defaults[Customizing OpenShift defaults].

[NOTE]
====
A major difference between the Loki Operator and Loki itself lies in inheritance handling. Loki only copies `default_resource_attributes_as_index_labels` to tenants by default, while the Loki Operator applies the entire global configuration to each tenant in `openshift-logging` mode.
====

Within `LokiStack`, attribute mapping configuration is managed through the `limits` setting:

[source,yaml]
----
# [...]
spec:
  limits:
    global:
      otlp: {} # <1> 
    tenants:
      application:
        otlp: {} # <2> 
----
<1> Global OTLP attribute configuration.
<2> OTLP attribute configuration for the `application` tenant within `openshift-logging` mode.

[NOTE]
====

> Both global and per-tenant OTLP configurations can map attributes to stream labels or structured metadata. At least _one stream-label_ is required to save a log entry to Loki storage, so ensure this configuration meets that requirement.

Reviewer: Is the emphasis really needed here? Also this is the first instance I see of "stream label" having a `-` in the middle:
```suggestion
Both global and per-tenant OTLP configurations can map attributes to stream labels or structured metadata. At least one stream label is required to save a log entry to Loki storage, so ensure this configuration meets that requirement.
```

====

Stream labels derive only from resource-level attributes, which the `LokiStack` resource structure reflects:

[source,yaml]
----
spec:
  limits:
    global:
      otlp:
        streamLabels:
          resourceAttributes:
          - name: "k8s.namespace.name"
          - name: "k8s.pod.name"
          - name: "k8s.container.name"
----

Structured metadata, in contrast, can be generated from resource, scope or log-level attributes:

[source,yaml]
----
# [...]
spec:
  limits:
    global:
      otlp:
        streamLabels:
          # [...]
        structuredMetadata:
          resourceAttributes:
          - name: "process.command_line"
          - name: "k8s\\.pod\\.labels\\..+"
            regex: true
          scopeAttributes:
          - name: "service.name"
          logAttributes:
          - name: "http.route"
----

[TIP]
====

> Use _regular expressions_ by setting `regex: true` for attributes names when mapping similar attributes in Loki. 

Reviewer: ```suggestion
Use regular expressions by setting `regex: true` for attributes names when mapping similar attributes in Loki. 
```

====

[IMPORTANT]
====
Avoid using regular expressions for stream labels, as this can increase data volume.
====

=== Customizing OpenShift defaults

In `openshift-logging` mode, certain attributes are required and cannot be removed from the configuration due to their role in OpenShift functions. Other attributes, labeled *recommended*, might be disabled if performance is impacted. For a complete attribute list, see xref:../../../observability/logging/logging-6.0/log6x-opentelemetry-data-model.adoc#attributes[OpenShift default attributes].

> When using the `openshift-logging` mode without custom attributes, you can achieve immediate compatibility with OpenShift tools. If additional attributes are needed as stream labels or structured metadata, use custom configuration. Custom configurations can *merge* with default configurations.

Reviewer: Is *merge* a label / UI element? Or can it be plaintext:
```suggestion
When using the `openshift-logging` mode without custom attributes, you can achieve immediate compatibility with OpenShift tools. If additional attributes are needed as stream labels or structured metadata, use custom configuration. Custom configurations can merge with default configurations.
```
Emphasis doesn't seem necessary, if that's what it was for.

Author: It be plaintext.

=== Removing recommended attributes

To reduce default attributes in `openshift-logging` mode, disable recommended attributes:

[source,yaml]
----
# [...]
spec:
  tenants:
    mode: openshift-logging
    openshift:
      otlp:
        disableRecommendedAttributes: true # <1>
----
<1> Set `disableRecommendedAttributes: true` to remove recommended attributes, which limits default attributes to the *required attributes*.

[NOTE]
====
This option is beneficial if the default attributes causes performance or storage issues. This setting might negatively impact query performance, as it removes default stream labels. You should pair this option with a custom attribute configuration to retain attributes essential for queries.
====

> == References

Reviewer: See repo guidelines for `== Additional resources` usage in an assembly.

:_mod-docs-content-type: ASSEMBLY
[id="log6x-configuring-lokistack-otlp"]
= OTLP data ingestion in Loki
include::_attributes/common-attributes.adoc[]
:context: log6x-configuring-lokistack-otlp

toc::[]

Loki 3.0 introduces an API endpoint using the OpenTelemetry Protocol (OTLP), providing a new way to ingest log entries into Loki. This endpoint complements the standard Push API, which is available in Loki from its initial release. Because OTLP is a standardized format not specifically designed for Loki, it requires additional configuration on Loki's side to map OpenTelemetry's data format to Loki's data model. OTLP lacks concepts such as _stream labels_ or _structured metadata_. Instead, OTLP provides metadata about log entries as *attributes*, grouped into three categories:

* Resource
* Scope
* Log

This allows metadata to be set for multiple entries simultaneously or individually as needed.

> include::modules/log6x-configuring-lokistack-otlp-data-ingestion.adoc[leveloffset=+1]

Author: I will discuss about this comment with Logging team and if relevant would like to cover as part of this PR: https://github.com/openshift/openshift-docs/pull/84717

* `otlp_config` in the link:https://grafana.com/docs/loki/latest/configure/#limits_config[`limits_config`]

By default, `default_resource_attributes_as_index_labels` maps certain resource attributes to stream labels within Loki. 

Since this distributor configuration only applies to resource-level attributes and generates stream labels as output, the `otlp_config` setting allows mapping resource, scope, or log-level attributes to structured metadata.

[NOTE]
====
Attributes that remain unmapped to either stream labels or structured metadata do not store in Loki.
====

The Loki Operator handles attribute configuration differently than Loki itself. Notably, the Loki Operator does not distinguish between `distributor` and `limits` configurations. Instead, the Loki Operator consolidates all attributes in the `limits` configuration. For more infomation on Operator-specific configuration options, see xref:../../../observability/logging/logging-6.0/log6x-configuring-lokistack-otlp.adoc#custom-attribute-mapping[Custom attribute mapping].

=== Loki Operator defaults

> When using the Loki Operator in `openshift-logging` mode, default attribute mappings follow OpenShift configurations. For more information on configuring custom attributes within `openshift-logging` mode, see xref:../../../observability/logging/logging-6.0/log6x-configuring-lokistack-otlp.adoc#customizing-openshift-defaults[Customizing OpenShift defaults].

Reviewer 2: This feels like duplication with the entry paragraph in the `Custom attribute mapping`. I suggest to choose one of the two to refer to the `Customizing OpenShift defaults`. Fact is that customizing attribute mapping is the logical next step to deviate from the defaults.

Author: To avoid duplication, removing the **Loki Operator defaults**  section seems better option to me.

=== Custom attribute mapping

> In `openshift-logging` mode, custom attribute mappings can be configured globally for all tenants or for individual tenants as needed. When custom mappings are defined, OpenShift defaults are bypassed. If default labels are still required, they must be included in the custom configuration. For more information, see xref:../../../observability/logging/logging-6.0/log6x-configuring-lokistack-otlp.adoc#customizing-openshift-defaults[Customizing OpenShift defaults].

Reviewer 2: ```suggestion
In `openshift-logging` mode, custom attribute mappings can be configured globally for all tenants or for individual tenants as needed. When custom mappings are defined, they are appended to the OpenShift defaults. If default recommended labels are not required, they can be disabled in the tenant configuration. For more information, see xref:../../../observability/logging/logging-6.0/log6x-configuring-lokistack-otlp.adoc#customizing-openshift-defaults[Customizing OpenShift defaults].
```

[NOTE]
====
A major difference between the Loki Operator and Loki itself lies in inheritance handling. Loki only copies `default_resource_attributes_as_index_labels` to tenants by default, while the Loki Operator applies the entire global configuration to each tenant in `openshift-logging` mode.
====

Within `LokiStack`, attribute mapping configuration is managed through the `limits` setting:

[source,yaml]
----
# [...]
spec:
  limits:
    global:
      otlp: {} # <1> 
    tenants:

>       openshift-logging:

Reviewer 2: Although we say tenant **mode** is `openshift-logging`, the three tenants are `application`, `infrastructure` and `audit`. I suggest to use `application` here for the example code.

By default, `default_resource_attributes_as_index_labels` maps certain resource attributes to stream labels within Loki. 

Since this distributor configuration only applies to resource-level attributes and generates stream labels as output, the `otlp_config` setting allows mapping resource, scope, or log-level attributes to structured metadata.

[NOTE]
====
Attributes that remain unmapped to either stream labels or structured metadata do not store in Loki.
====

The Loki Operator handles attribute configuration differently than Loki itself. Notably, the Operator does not distinguish between `distributor` and `limits` configurations. Instead, the Loki Operator consolidates all attributes in the `limits` configuration. For more infomation on Operator-specific configuration options, see xref:../../../observability/logging/logging-6.0/log6x-configuring-lokistack-otlp.adoc#custom-attribute-mapping[Custom attribute mapping].

=== Loki Operator defaults

When using the Loki Operator, default attribute mappings vary based on the tenancy mode used for the `LokiStack`:

* `static` and `dynamic` modes use Grafana defaults
* `openshift-logging` mode uses OpenShift defaults

> 

Reviewer 2: :warning: As with the entire page being a wrap-up from the upstream docs, we are interested our docs only about the tenant mode `openshift-logging`. Same with the `Custom attribute mapping`, we need the `Customizing OpenShift defaults` in our docs.
